src/commands/publish/upload_form/mod.rs

```diff
@@ -0,0 +1,145 @@
+use crate::settings::metadata::Metadata;
+use crate::settings::project::{Project, ProjectType};
+
+use project_assets::ProjectAssets;
+use script::Script;
+use wasm_module::WasmModule;
+
+use super::{krate, Package};
+
+pub fn build_script_upload_form(project: &Project) -> Result<Form, failure::Error> {
+    let project_type = &project.project_type;
+    match project_type {
+        ProjectType::Rust => {
+            info!("Rust project detected. Publishing...");
+            let name = krate::Krate::new("./")?.name.replace("-", "_");
+            // TODO: move into build?
+            build_generated_dir()?;
+            concat_js(&name)?;
+
+            let wasm_module = WasmModule {
+                path: format!("./pkg/{}_bg.wasm", name).to_string(),
+                filename: "wasm".to_string(),
+                binding: "wasmprogram".to_string(),
+            };
+
+            let script = Script {
+                path: "./worker/generated/script.js".to_string(),
+            };
+
+            let assets = ProjectAssets {
+                script,
+                wasm_modules: vec![wasm_module],
+            };
+
+            build_form(&assets)
+        }
+        ProjectType::JavaScript => {
+            info!("JavaScript project detected. Publishing...");
+            let package = Package::new("./")?;
+
+            let script = Script {
+                path: package.main()?,
+            };
+
+            let assets = ProjectAssets {
+                script,
+                wasm_modules: Vec::new(),
+            };
+
+            build_form(&assets)
+        }
+        ProjectType::Webpack => {
+            info!("Webpack project detected. Publishing...");
+            // FIXME(sven): shouldn't new
+            let bundle = Bundle::new();
+
+            let script = Script {
+                path: bundle.script_path(),
+            };
+
+            let mut wasm_modules = Vec::new();
+
+            if bundle.has_wasm() {
+                let wasm_module = WasmModule {
+                    path: bundle.wasm_path(),
+                    filename: bundle.get_wasm_binding(),
+                    binding: bundle.get_wasm_binding(),
+                };
+                wasm_modules.push(wasm_module)
+            }
+
+            let assets = ProjectAssets {
+                script,
+                wasm_modules,
+            };
+
+            build_form(&assets)
+        }
+    }
+}
+
+fn build_form(assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    let mut form = Form::new();
+
+    form = add_files(form, assets)?;
+    form = add_metadata(form, assets)?;
+
+    Ok(form)
+}
+
+fn add_files(mut form: Form, assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    for file in assets.files() {
+        form = form.file(file.name, file.path)?;
+    }
+
+    Ok(form)
+}
+
+fn add_metadata(mut form: Form, assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    let metadata_json = serde_json::json!(&Metadata {
```

Comment:
use of the macro here may be inappropriate, idk...

Reply:
should be fine, might make errors weird but a lot of the values here are hardcoded so not likely to throw